forge/src/multi_runner.rs

```diff
@@ -191,13 +191,51 @@ pub struct MultiContractRunner {
 }
 
 impl MultiContractRunner {
-    pub fn test(
+    pub fn test_stream(
         &mut self,
         filter: &(impl TestFilter + Send + Sync),
+        stream_result: impl Fn(String, BTreeMap<String, TestResult>) + Sync,
     ) -> Result<BTreeMap<String, BTreeMap<String, TestResult>>> {
-        // TODO: Convert to iterator, ideally parallel one?
         let contracts = std::mem::take(&mut self.contracts);
+        let vicinity = self.evm_opts.vicinity()?;
+        let backend = self.evm_opts.backend(&vicinity)?;
+        let source_paths = self.source_paths.clone();
 
+        let results = contracts
+            .par_iter()
+            .filter(|(name, _)| filter.matches_path(source_paths.get(*name).unwrap()))
+            .filter(|(name, _)| filter.matches_contract(name))
+            .map(|(name, (abi, deploy_code, libs))| {
+                // unavoidable duplication here?
+                let result = match backend {
+                    BackendKind::Simple(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), libs, filter)?
+                    }
+                    BackendKind::Shared(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), libs, filter)?
+                    }
+                };
+                Ok((name.clone(), result))
+            })
+            .filter_map(|x: Result<_>| x.ok())
+            .filter_map(
+                |(name, result)| if result.is_empty() { None } else { Some((name, result)) },
+            )
+            .map(|(name, result)| {
+                stream_result(name.clone(), result.clone());
+                (name, result)
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        self.contracts = contracts;
+        Ok(results)
+    }
+
+    pub fn test(
+        &mut self,
+        filter: &(impl TestFilter + Send + Sync),
+    ) -> Result<BTreeMap<String, BTreeMap<String, TestResult>>> {
+        let contracts = std::mem::take(&mut self.contracts);
```

Comment:
Perhaps we can deduplicate by making the streaming function optional and have this be a wrapper around calling `test_stream` without the streaming function

Reply:
Not sure what's idiomatic for an optional closure that is Sync. Maybe `Option<Box<dyn T + Sync>>`?

An alternative would be to get rid of `test` and give `test_stream` a dummy closure where necessary.

Reply:
Or you could just pass a dummy closure to `test_stream` in `test` :smile:

Reply:
`Option<Box<dyn _ + Send + Sync>>` worked fine. I was expecting it to be more problematic